### ISSUE TITLE
Ability to add new tokens and keep existing token embeddings frozen?

### ISSUE THREAD

**Issue by DreamGenX** [huggingface/peft#1462]
### Feature request

Today, when you add new tokens to the vocabulary (e.g. `<|im_start|>` and `<|im_end|>`), you need to also add `embed_tokens` and `lm_head` to the `modules_to_save` kwarg. This, as far as I can tell, unfreezes all token embeddings.

Would it be possible to unfreeze only the newly added tokens?

### Motivation

This would allow us to expand the vocabulary, and not fine tune too many extra params.

### Your contribution

Seeking feedback

**Comment by BenjaminBossan** [comment]
> you need to also add `embed_tokens` and `lm_head` to the `tokens_to_save` kwarg

I assume you mean `modules_to_save` and not `tokens_to_save`. Indeed, this makes the whole embedding layer learnable. AFAIK, there is no easy way in PyTorch to make only a part of the parameters learnable and freeze the rest. This would probably require a custom module class to work.

A possible solution to avoid having to update the possibly huge embedding weights is to apply LoRA (assuming you use LoRA) to that layer. More on this can be found in the docs [here](https://huggingface.co/docs/peft/developer_guides/troubleshooting#extending-the-vocabulary).

**Comment by DreamGenX** [comment]
> I assume you mean `modules_to_save` and not `tokens_to_save`

Oh yes, updated.

> A possible solution to avoid having to update the possibly huge embedding weights is to apply LoRA (assuming you use LoRA) to that layer.

Yeah, that would certainly help reduce the parameter count.

**Comment by marcusinthesky** [comment]
I wonder if this could be done, using [sparse matrices](https://pytorch.org/docs/stable/sparse.html#supported-operations)? You could initial LoRA's A matrix as a sparse matrix, or `modules_to_save` as a sparse matrix. My only concern would by [safetensor's support](https://github.com/huggingface/safetensors/blob/b947b59079a6197d7930dfb535818ac4896113e8/bindings/python/py_src/safetensors/torch.py#L397).

**Comment by BenjaminBossan** [comment]
@marcusinthesky I'm not sure how sparse matrices would help here, could you please elaborate? LoRA weights are already sparse in the sense that they're a low rank approximation of a much bigger matrix. When it comes to `modules_to_save`, we create a copy of the original module, so the parameter type has to be the same.

**Comment by marcusinthesky** [comment]
The actual A and B matrices that we initialize are dense.  They are not using Pytorch Sparse Matrix structure like COO etc. To update the embeddings at particular token positions it seems like we have three options:
1. We add a gradient hook to zero ylthe gradient at particular indices... but we still have to store the full embedding matrix or delta,
2.  We store only the deltas for the tokens we want to update, then either 'upcast' them to the full NxK embedding matrix size or index down the embedding weight matrix and add, or
3. We initialize a Pytorch Sparse Matrix type Delta and then add then delta and store that sparse delta directly

**Comment by BenjaminBossan** [comment]
> The actual A and B matrices that we initialize are dense

Yes, they are dense, what I mean is that they cannot really be made any sparser, as they're already a low-rank representation.

> 3\. We initialize a Pytorch Sparse Matrix type Delta and then add then delta and store that sparse delta directly

Do you have a code snippet to illustrate this?

**Comment by marcusinthesky** [comment]
> Yes, they are dense, what I mean is that they cannot really be made any sparser, as they're already a low-rank representation.
So, like with my example above, we LoRA only a sparse subset of tokens, like:
```
import torch

N = 1000 # this is the number of tokens in the vocabulary
K = 512 # this is the dimension of the embeddings

T = 10 # number of tokens we want to let trainable
R = 64 # lora rank

# we initial a random pytorch matrix which is NxK
W = torch.rand(N, K) # this is the original embedding weight

r = torch.randint(0, N, (T, )) # we randomly select which tokens are have trainable delta's
c = torch.arange(0, R)

# we now need to tile and repeat the r and c to get the indices
indices = torch.stack([r.repeat(R), c.repeat(T, 1).t().reshape(-1)], dim=1).T

# we also need to initialize the values
values = torch.rand(T*R, requires_grad=True)

# we create the sparse tensor
A = torch.sparse_coo_tensor(indices=indices, values=values, size=(N, R))
B = torch.rand(R, K, requires_grad=True)


D = A @ B

W_prime = W + D # this is the new embedding weight, which is the original weight plus the trainable delta's at tokens r

# fake loss function
loss = (W_prime**2).sum()
loss.backward()

# Inspect the gradients of the values of D's values
print("Gradients with respect to D:")
print(values.grad)
```

**Comment by marcusinthesky** [comment]
I've created a draft PR, I do need to go through how tests and docs are written in peft still, this was a first crack. I think one application of this Tuner may be in loading new tokens, I am just nervous around baking that into the tuner, as that requires aligning the Tokenizer. I think we should assume that new token's have been created and we just need to update them. Happy for feedback here.
